In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Load dataset
data = load_breast_cancer()
X = pd.DataFrame(data.data, columns=data.feature_names)
y = pd.Series(data.target)

# Check for missing values
print(X.isnull().sum().sum())

# Feature scaling
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)


0


Explanation

Missing values: The dataset has no missing values, so no imputation is needed.

Feature scaling: Standardization (StandardScaler) ensures all features have mean 0 and variance 1. This is crucial for algorithms like SVM and k-NN, which are sensitive to feature magnitudes.

Train-test split: We split data into training (80%) and testing (20%) to evaluate performance fairly.

In [2]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

# Initialize models
models = {
    "Logistic Regression": LogisticRegression(max_iter=500),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "Random Forest": RandomForestClassifier(random_state=42),
    "SVM": SVC(),
    "k-NN": KNeighborsClassifier()
}

# Train and evaluate
results = {}
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    results[name] = accuracy_score(y_test, y_pred)

results


{'Logistic Regression': 0.9736842105263158,
 'Decision Tree': 0.9473684210526315,
 'Random Forest': 0.9649122807017544,
 'SVM': 0.9736842105263158,
 'k-NN': 0.9473684210526315}

Algorithm Descriptions

Logistic Regression: Models probability using a logistic function. Suitable for binary classification like breast cancer (malignant vs benign).

Decision Tree: Splits data into branches based on feature thresholds. Easy to interpret but prone to overfitting.

Random Forest: Ensemble of decision trees; reduces overfitting and improves accuracy.

SVM: Finds the optimal hyperplane to separate classes. Works well with high-dimensional data like this dataset.

k-NN: Classifies based on majority vote of nearest neighbors. Simple but sensitive to scaling and choice of k.

In [3]:
# Display results
for model, acc in results.items():
    print(f"{model}: {acc:.4f}")


Logistic Regression: 0.9737
Decision Tree: 0.9474
Random Forest: 0.9649
SVM: 0.9737
k-NN: 0.9474


Comparison

Best performers: Logistic Regression and SVM (both at 97.37%).

This makes sense because the dataset is linearly separable to a large extent, so both linear models (Logistic Regression, SVM with linear kernel) excel.

Worst performers: Decision Tree and k-NN (both at 94.74%).

Decision Trees can overfit small variations in the data.

k-NN is sensitive to scaling and the choice of k; while scaling helped, it still lags behind ensemble and linear models.

Middle performer: Random Forest (96.49%) — strong, but slightly behind the linear models here.